diff -urN a/setup.py b/setup.py
--- a/setup.py	2013-04-21 22:07:58.004660384 +0100
+++ b/setup.py	2013-04-21 22:07:58.398663375 +0100
@@ -1004,7 +1004,8 @@
                 for p in ['msvcrtmodule.c']]) )
 
             exts.append( Extension('_msi', [os.path.join(pc_srcdir, p)
-                for p in ['_msi.c']]) )
+                for p in ['_msi.c']],
+                libraries=['msi','cabinet','rpcrt4']) ) # To link with lib(msi|cabinet|rpcrt4).a
 
             exts.append( Extension('_subprocess', [os.path.join(pc_srcdir, p)
                 for p in ['_subprocess.c']]) )
