diff -urN a/setup.py b/setup.py
--- a/setup.py	2013-04-21 22:08:03.738703907 +0100
+++ b/setup.py	2013-04-21 22:08:04.128706867 +0100
@@ -1369,7 +1369,7 @@
 
         dbm_order = ['gdbm']
         # The standard Unix dbm module:
-        if host_platform not in ['cygwin']:
+        if host_platform not in ['cygwin', 'mingw', 'win32']:
             config_args = [arg.strip("'")
                            for arg in sysconfig.get_config_var("CONFIG_ARGS").split()]
             dbm_args = [arg for arg in config_args
