diff -urN a/setup.py b/setup.py
--- a/setup.py	2013-05-02 00:44:52.984562711 +0100
+++ b/setup.py	2013-05-02 00:44:53.383566007 +0100
@@ -484,6 +484,19 @@
             if not with_build_sysroot_specified:
                 add_dir_to_list(self.compiler.library_dirs, '/usr/local/lib')
                 add_dir_to_list(self.compiler.include_dirs, '/usr/local/include')
+
+            # Because tcltk is built shared, we build it directly into the
+            # same prefix as where we're building Python. This requires
+            # that I add the prefix folder to library_dirs and include_dirs.
+            # Instead, should build it to the same prefix used for other
+            # libraries (i.e. the one specified with --with-build-sysroot),
+            # BUILD_SYSROOT, then run 'DESTDIR=$BUILD_SYSROOT make install'
+            # for both tcl and tk.
+            # This allows any lib to be --prefix'd using Python's --prefix
+            norm_prefix = os.path.normpath(sys.prefix)
+            add_dir_to_list(self.compiler.library_dirs, norm_prefix + '/lib')
+            add_dir_to_list(self.compiler.include_dirs, norm_prefix + '/include')
+
             self.add_multiarch_paths()
 
         # Add paths specified in the environment variables LDFLAGS and
