diff -urN a/setup.py b/setup.py
--- a/setup.py	2013-12-11 16:31:09.481613757 +0000
+++ b/setup.py	2013-12-11 16:31:09.771618253 +0000
@@ -2285,8 +2285,9 @@
                 newfilename = filename + fullversion
             else:
                 newfilename = filename + minoronly
-            log.info('renaming {} to {}'.format(filename, newfilename))
-            os.rename(filename, newfilename)
+            if not os.path.exists(newfilename):
+                log.info('renaming {} to {}'.format(filename, newfilename))
+                os.rename(filename, newfilename)
             newoutfiles.append(newfilename)
             if filename in updated_files:
                 newupdated_files.append(newfilename)
