diff -urN a/setup.py b/setup.py
--- a/setup.py	2013-12-11 14:57:41.918227668 +0000
+++ b/setup.py	2013-12-11 14:57:42.201565358 +0000
@@ -1664,7 +1664,7 @@
 ##         ext = Extension('xx', ['xxmodule.c'])
 ##         self.extensions.append(ext)
 
-        if 'd' not in sys.abiflags:
+        if 'd' not in sys.abiflags and not COMPILED_WITH_PYDEBUG:
             ext = Extension('xxlimited', ['xxlimited.c'],
                             define_macros=[('Py_LIMITED_API', 1)])
             self.extensions.append(ext)
