diff -urN a/configure.ac b/configure.ac
--- a/configure.ac	2013-12-11 00:08:55.625908900 +0000
+++ b/configure.ac	2013-12-11 00:09:02.412297100 +0000
@@ -2148,14 +2148,14 @@
 # -- usually .so, .sl on HP-UX, .dll on Cygwin
 AC_MSG_CHECKING(the extension of shared libraries)
 if test -z "$SHLIB_SUFFIX"; then
-	case $ac_sys_system in
+	case $MACHDEP in
 	hp*|HP*)
 		case `uname -m` in
 			ia64) SHLIB_SUFFIX=.so;;
 	  		*)    SHLIB_SUFFIX=.sl;;
 		esac
 		;;
-	CYGWIN*)   SHLIB_SUFFIX=.dll;;
+	CYGWIN*|win*)   SHLIB_SUFFIX=.dll;;
 	*)	   SHLIB_SUFFIX=.so;;
 	esac
 fi
