diff -urN a/Modules/Setup.config.in b/Modules/Setup.config.in
--- a/Modules/Setup.config.in	2013-12-26 17:48:43.820313800 +0000
+++ b/Modules/Setup.config.in	2013-12-26 17:48:46.819485400 +0000
@@ -15,6 +15,8 @@
 # On win32 host(mingw build in MSYS environment) show that site.py
 # fail to load if some modules are not build-in:
 @BUILDIN_WIN32_MODULE@winreg ../PC/winreg.c
+@BUILDIN_WIN32_MODULE@time timemodule.c
+@BUILDIN_WIN32_MODULE@msvcrt ../PC/msvcrtmodule.c
 
 
 # The rest of the modules previously listed in this file are built
